### ISSUE TITLE
Discussion: Regrading Streaming and Talkback (Two Way Audio)

### ISSUE THREAD

**Issue by AngellusMortis** [briis/pyunifiprotect#87]
So thinking about trying to implement a WebRTC backend for a UFP camera or a UDP stream connection for Two Way Audio, where do we want to implement that? In `pyunifiprotect` or `unifiprotect`? I know HA already has some "streaming" logic in it, so it _may_ make sense to put it to the HA integration, but then then if anyone wants to use the library outside of HA, they would need to implement their own solution. 

If we do implement in the library, it will give the library a hard requirement of needing ffmpeg to manage the streams/transcodes/etc.

As for interface, I was thinking of something like this:

WebRTC:
```python
def create_webrtc_stream(camera):
    ffmpeg_url = create_ffmpeg_transcode() # will spawn ffmpeg to create a video stream that transcodes the AAC audio into OPUS for WebRTC
    # some kind of singleton to prevent multiple streams for a single camera
    
    url = create_webrtc_backend(ffmpeg_url) # create WebRTC stream with aiortc
    return {url: url}

stream = create_webrtc_stream(camera)
stream.url # + whatever else HA needs to play the Web RTC stream 
stream.close() # maybe an automatic time out instead to close the session after X time of unuse? To keep from always spinning up
```

Playing audio (media player) / Two Way Audio:

```python
def create_audio_stream(camera):
     ffmpeg_url = create_ffmpeg_stream() # Create a UDP connection to camera: https://github.com/hjdhjd/homebridge-unifi-protect/blob/f236df00a4a47ceb8ae5ca3a84305ff786fc4f23/src/protect-stream.ts#L487
     # again singleton of some kind

     retturn ffmpeg_url

steam = create_audio_stream(camera)
stream.write(raw_audio)
stream.close()  # maybe an automatic time out instead to close the session after X time of unuse? To keep from always spinning up

# some kind of wrapper class to one-off play audio to the camera to support the media_player interface
```

The Two Way Audio stuff would be "implemented, but unused" until HA got better two way audio support, but this is just discussing now where we would want it.

**Comment by bdraco** [comment]
Probably implement in pyunifiprotect but let the integration pass in the path to the ffmpeg binary.  There is an example of this in the HomeKit integration in type_cameras.py

**Comment by AngellusMortis** [comment]
That would certainly work. Does Home Assistant already have [pyav](https://pypi.org/project/av/) and [pylibsrtp](https://pypi.org/project/pylibsrtp/)? Those looks like the nasty requirements that need a ton of native deps to work.

**Comment by bdraco** [comment]
The stream integration uses a forked pyav. AFAIK, the maintainer of pyav only created it for semi personal use and it grew much larger than expected and there isn't much motivation to continue to extend it so HA ended up with a forked version.  That might change but seems to be the current status quo

**Comment by AngellusMortis** [comment]
It looks like the fork will probably work since `pip list` shows `av`. Hopefully the fork is compatible.

`pylibsrtp` does not seem to exist and it looks like my HA install does not have needed files to build it. It needs `libsrtp2-dev` (Debian base) or `srtp` (MacOS). Is there a good point for getting that added?

**Comment by bdraco** [comment]
I think we would need to merge the main integration to core first. That's probably doable rather quickly once we fix some of the standards issues

**Comment by AngellusMortis** [comment]
Okay. Anything I implement will just assume it is there then. Integrating with core was planned after I get the new API client stuff done (MVP needed: need to still be able to update UFP objects, prefer to get talkback stuff done) and then update the integration to use the new API client.

**Comment by bdraco** [comment]
I'll do a run through for standards some time later this week once my flight schedule calms down and see if I can take care of anything that will hold up a merge to core

**Comment by briis** [comment]
Agree that it is about time to merge this in to core. Looking forward to hear what needs changing, before that can be done.

**Comment by AngellusMortis** [comment]
Great news! I finally figured out talkback.

If you get the raw JSON for your camera (either via `/proxy/protect/api/bootstrap` or `/proxy/protect/api/cameras/{id}`), there is a `talkbackSettings` obj.

I _think_ these settings are always the same, but references just in case they are not:

```json
{
...
    "talkbackSettings": {
        "typeFmt": "aac",
        "typeIn": "serverudp",
        "bindAddr": "0.0.0.0",
        "bindPort": 7004,
        "filterAddr": "",
        "filterPort": 0,
        "channels": 1,
        "samplingRate": 22050,
        "bitsPerSample": 16,
        "quality": 100
    },
...
}
```

Using ffmpeg, you can then play an audio stream to your camera. The format is:

```bash
ffmpeg -loglevel error -hide_banner {input_args} -vn -acodec {talkbackSettings.typeFmt} -ac {talkbackSettings.channels} -ar {talkbackSettings.sampleingRate} -bits_per_raw_sample {talkbackSettings.bitsPerSample} -map 0:a -aq {talkbackSettings.quality} -f adts udp://{host}:{talkbackSettings.bindPort}
```

So if you want to play a MP3 file file named `test.mp3` to your camera  at `192.168.50.10`:

```bash
ffmpeg -loglevel error -hide_banner -i test.mp3 -vn -acodec aac -ac 1 -ar 22050 -bits_per_raw_sample 16 -map 0:a -aq 100 -f adts udp://192.168.50.10:7004
```

**Comment by briis** [comment]
Woow, cool. Besides implementing this we now just need a smart way to make this available in HA UI.

**Comment by AngellusMortis** [comment]
Well, full Two Way Talk is probably a no go. At least right now, but it does mean we can implement a `media_player` entity.

**Comment by briis** [comment]
Yes we can, and I could see a use scenario in sending a pre-recorded audio message to a camera or a Doorbell.

**Comment by AngellusMortis** [comment]
Or TTS.

**Comment by briis** [comment]
> Or TTS.

Oh yes, completely forgot about that - Even cooler.

**Comment by briis** [comment]
> I think we would need to merge the main integration to core first

I setup the Core Cev environment, and added unifiprotect to it. With some changes, it will now run fine in core. So I hope it will be a minor task to move it. There are still some points on the [Development Checklist](https://developers.home-assistant.io/docs/development_checklist) I have not done, as I am not sure what they all do. So when we are ready and Protect 1.20 is released, I hope you can run through this also @bdraco. 
I have started on the documentation, so that should be ready when the code is ready.

**Comment by AngellusMortis** [comment]
I do not think we should move into core until we get some other stuff done since integrating into core will slowdown the interval in which we can release updates. Stuff I was thinking about:

* Start using `ProtectApiClient`
* Implement talkback / `media_player` entities
* Add a few new entities for cameras:
    * Sensor for last motion event (with HA proxied link to thumbnail)
    * Sensor for last smart detect event (with HA proxied link to thumbnail)
    * Binary sensor for Is Dark, maybe? (based on troubles user had in unifiprotect integration issue)
* (Maybe, but can probably wait) Implement media browner / playback for Events (basically re-implement the Events page from Protect in HA)

That should knock out the bulk of the issues and major changes needed for the next month or so.

**Comment by briis** [comment]
I do agree with you Christopher - Except for the last bullet, we should implement this before moving to core.
* Bullet 1: I will wait for you to be done, and then I can start migrating the Unfiprotect Integration. Should most likely be after the 0.10.0 release.
* Bullet 2: If you can write the function in pyunifiprotect, I can do the media player pretty fast. Maybe even for 0.10.0.
* Bullet 3: The two first should wait til 0.11 release, as we need to sort out the timing of thumbnails. The is_dark, I can do now.
* Bullet 4: Yes, this looks a bit more complex

**Comment by AngellusMortis** [comment]
I got a working prototype of talkback working, but it is going to require the new API client. So it will have to wait for the 0.11 release. After https://github.com/briis/pyunifiprotect/pull/89 gets merged, I will post the PR.

**Comment by AngellusMortis** [comment]
Also, I just discovered via the [forums](https://community.home-assistant.io/t/webrtc-or-lhls-support-for-camera-stream-components/344909/5) that LL HLS support _was_ sneakily added to HA without anything in the change logs. Add `ll_hls: True` to your `stream:` component config in your `configuration.yaml` to enable it. It drops the latency for my Protect cameras drop to ~3 seconds which is signfanctly better then before.

**Comment by bdraco** [comment]
> Also, I just discovered via the [forums](https://community.home-assistant.io/t/webrtc-or-lhls-support-for-camera-stream-components/344909/5) that LL HLS support _was_ sneakily added to HA without anything in the change logs. Add `ll_hls: True` to your `stream:` component config in your `configuration.yaml` to enable it. It drops the latency for my Protect cameras drop to ~3 seconds which is signfanctly better then before.
> 
> 
> 
> 

It's kinda in testing at the moment 😉

**Comment by briis** [comment]
Woow, yes the ll_hls really makes an impact.

**Comment by bdraco** [comment]
I ran through the unifiprotect HA code and identified a few standards issues.  I'm out of time as I've got to run to the airport so I only got though 3 of the files.

- `config_flow.py`

1. Options need to be moved to an options flow https://developers.home-assistant.io/docs/config_entries_options_flow_handler/ since config flows are only supposed to collect the bare minimum amount of information to be able to setup the device.  The goal is to make getting started as easy as possible and let the user configure options later.
```
                CONF_DISABLE_RTSP: user_input.get(CONF_DISABLE_RTSP),
                CONF_DOORBELL_TEXT: user_input.get(CONF_DOORBELL_TEXT),
                CONF_SNAPSHOT_DIRECT: user_input.get(CONF_SNAPSHOT_DIRECT),
                CONF_SCAN_INTERVAL: DEFAULT_SCAN_INTERVAL,
```

2. 100% test coverage is required for the config flow to merge

- `__init__.py` 

1. Drop `async_setup` as its no longer needed

2.

```python3
    data["stop_event_listener"] = hass.bus.async_listen_once(
        EVENT_HOMEASSISTANT_STOP, _async_stop_protect_data
    )
```
We can now get rid of `stop_event_listener` and wrap `hass.bus.async_listen_once` in `entry.async_on_unload`

We can now get rid of `update_listener` and wrap `entry.add_update_listener(_async_options_updated)` in `entry.async_on_unload` and then delete
```
        data["update_listener"]()
        data["stop_event_listener"]()
``` from `async_unload_entry`

3.

Replace
```python3
    for platform in UNIFI_PROTECT_PLATFORMS:
        hass.async_create_task(
            hass.config_entries.async_forward_entry_setup(entry, platform)
        )
```

with

`hass.config_entries.async_setup_platforms(entry, UNIFI_PROTECT_PLATFORMS)`

4. 
Replace
```python3
    unload_ok = all(
        await asyncio.gather(
            *[
                hass.config_entries.async_forward_entry_unload(entry, component)
                for component in UNIFI_PROTECT_PLATFORMS
            ]
        )
    )
```
with

`python3
unload_ok = await hass.config_entries.async_unload_platforms(entry, UNIFI_PROTECT_PLATFORMS)
```

- `binary_sensor.py`

1. Don't pass in `hass` or set `self._hass` as `self.hass` will be set when the entity is added to Home Assistant in `async_added_to_hass`

2. Firing events from an entity is not allowed as its duplicate functionality and writes rows to the database for the state change event, and also the event being fired.  The normal advice is to implement these as device triggers instead.

Note that 3 & 4 are kinda moot because of 2

3. Firing events from a property is not allowed `is_on` calls `self._hass.bus.fire`.

Modify `entity.py`'s to call another function, maybe `self.async_process_update` instead of `self.async_write_ha_state` and have `async_process_update` call `self.async_write_ha_state`, then in `binary_sensor.py`, override the function to fire the event and call `super().async_process_update()` method.

```python3

    async def async_added_to_hass(self):
        """When entity is added to hass."""
        self.async_on_remove(
            self.protect_data.async_subscribe_device_id(
                self._device_id, self.async_write_ha_state
            )
        )

```

4. `is_on` is using the sync api in async code `self._hass.bus.async_fire` should be used instead here (but it should be moved to another function per 2)

**Comment by briis** [comment]
Thanks Nick,
I will work on these. Have a safe journey.

**Comment by AngellusMortis** [comment]
So with getting feature parity with `UpvServer` for `ProtectApiClient` and adding the talkback stuff, what do you guys think about making an actual 1.0 version of `pyunifiprotect`? Seems like a good enough time as any with the breaking changes. 

I was thinking we could add deprecation warnings for `UpvServer` and then remove it in a 1.1 version or something.

**Comment by briis** [comment]
With all the changes, and a feature complete `pyunifiprotect` I agree this is a good time. I know it will never be real feature complete, as we and Ubiquiti will come up with new stuff, but with all the great work you guys did I think this is a V1.0.

> I was thinking we could add deprecation warnings for UpvServer and then remove it in a 1.1 version or something.

Agree

Once V1.0 is done - I will branch out a version of `unifiprotect` and make sure all works with `ProtectApiClient`. Then let us see how far we get with that before Protect V1.20 is released. and make the decision at that point.

**Comment by AngellusMortis** [comment]
Everything should be "done". I can add some deprecation warning and do another once over for the doc strings if you want before you actually cut the 1.0 version.

**Comment by briis** [comment]
So basically I would import` ProtectApiClient` instead of `UpvServer` and then all (or most) of the function calls would be the same? Is that correct?

**Comment by briis** [comment]
I will build the V1.0 now.

Also Chris: What are the functions named that you have setup for the use with the Media Player? There is `play_audio` but was there also a separate function to control the Volume?

**Comment by AngellusMortis** [comment]
No, it is definitely going to be a bit more complicated then that. Run down of the differences:

* `update` still exists on both and pretty much functions the same way
* Instead of `_processed_devices` and many of the `get_` methods you called before you get the data from UFP, basically everything lives in the `.bootstrap` property on the API client. It is a real time record of the state of the NVR / UFP (it is updated automatically by the Websocket). Bootstrap and everything under it is fully type hinted so that is the best way to figure out how to convert calls for devices/info from `UpvServer` to `ProtectApiClient`, but here are some examples:
    * `_get_unique_id()` -> `.bootstrap.nvr.mac`
    * `_get_server_info()` -> `{"server_name": .boostrap.nvr.name, ...}`
    * `get_snapshot_image()` -> `.boostrap.cameras[camera_id].get_snapshot()`
    * `set_camera_ir() -> `.boostrap.cameras[camera_id].set_ir_led_model()`
* `subscribe_websocket` still exists, but now instead of giving you a dictionary of the event, it gives you a `WSSubscriptionMessage` (again full type hints)

**Comment by briis** [comment]
OK, thanks. I might need some help when I start doing this, especially on the Websocket part.

**Comment by AngellusMortis** [comment]
`camera.play_audio` to play audio and `camera.set_speaker_volume` to adjust the volume. Calling a `.set_` method on a device (camera / light /etc.) will automatically update that specific device and then send the data _back_ to UFP to update it. I made a nice "simplified" PR for you to see how adding new `.set_` methods work since that is something I would except you to want to do potentially often. https://github.com/briis/pyunifiprotect/pull/90

**Comment by briis** [comment]
V1.0.0 is uploaded to Pypi.org. 
Thank you both - I don't understand all the code yet, but it is truly awesome

**Comment by AngellusMortis** [comment]
Darn. I tried to beat you to it, but I merged some changes for some updates for docstrings. Since we have the new stuff in there, I was going to post details of the API client on the Ubiquiti Discord.

**Comment by briis** [comment]
:) I will do a version 1.0.1. Is it OK to do now?

**Comment by AngellusMortis** [comment]
Yep. All done.

**Comment by AngellusMortis** [comment]
Regrading Thumbnail images (and eventually MP4 clips for events as well)...

You need to make a view for the `unifiprotect` integration to proxy the HTTP calls from HA to UFP. When I was testing playing audio, I noticed the TTS component does it, so here is an example from the TTS component:

Make a "view" class: https://github.com/home-assistant/core/blob/dev/homeassistant/components/tts/__init__.py#L623

Inheriting from `HomeAssistantView`: https://github.com/home-assistant/core/blob/dev/homeassistant/components/http/view.py#L31

Then register it in your `async_setup`: https://github.com/home-assistant/core/blob/b1d49b3b663a65b4ae1b149b4e558bba0e64c46e/homeassistant/components/tts/__init__.py#L137

The view would basically just be something like `/api/ufp/thumbnail/{thumbnail_id}` and then it would just be a wrapper around `ProtectApiClient.get_event_thumbnail`. Then you can add URLs as attributes to events for that use that view. No more service calls to download images needed!

**Comment by briis** [comment]
Thanks. I will get back to you on this, once I am ready for that. Right now, I just to try to clean-up the code based on the recommendations from @bdraco, and then comes the migration to the `ProtectApiClient`.

**Comment by AngellusMortis** [comment]
No worries, I was just thinking about it and wanted to look it up before I forgot.

**Comment by briis** [comment]
@bdraco As you know I am working my way through the first suggestions, but I am struggling with the suggestions for the Binary Sensor. I know what you want me to do, but I am not sure how to do it. Is there a good example somewhere?

**Comment by bdraco** [comment]
> @bdraco As you know I am working my way through the first suggestions, but I am struggling with the suggestions for the Binary Sensor. I know what you want me to do, but I am not sure how to do it. Is there a good example somewhere?

I think the easiest solution is to remove the `hass.bus.fire` calls.  Users can already watch for the entity to change state to accomplish the same thing. It will also reduce the amount of rows that the integration creates in the database by almost half since thats where most of the data comes from.

**Comment by AngellusMortis** [comment]
I assume `hass.bus.fire` is to fire events? Yeah, I do not use events in my system. Granted, I use NodeRed instead of the HA automation system, but I just monitor the state/attribute changes I care about directly.

**Comment by bdraco** [comment]
> I assume `hass.bus.fire` is to fire events?

Yes.  Since state changes already generate events, core doesn't allow firing events from entities since it results in duplicate data in the database and extra overhead

**Comment by briis** [comment]
The reason I added this was that some users requested events for motion triggers, so removing this will create a breaking change. I was already looking at device_trigger support, but was not completely sure how to implement that, even after reading the developer docs.
But I am fine with removing for now.

**Comment by AngellusMortis** [comment]
Since it is something that needs to be changed for core, I would get the code ready for removal, but do _not_ remove it just yet. Then when we release 0.11, which will likely be the last HACS release, deprecate it and tell users it will be removed in the next release. Then when we merge for core, we just remove it then.

**Comment by bdraco** [comment]
If the wifi works in flight today, I will go through the other platforms and look for issues that would block merging to core

**Comment by briis** [comment]
Thanks Nick, appreciate it. Have a good flight.

**Comment by AngellusMortis** [comment]
Added tags/releases to `pyunifiprotect` repo.

**Comment by bdraco** [comment]
I ran through the rest of the platforms.  The stand out is to use `XXXEntityDescription` to avoid the if/if/if blocks.

- `__init__.py`

Only store data in hass.data after the integration has passed all the places it could retry via `ConfigEntryNotReady`

Move
```python3
hass.data.setdefault(DOMAIN, {})[entry.entry_id] = protectserver
```

to  after the last `ConfigEntryNotReady`

Likely  (untested) it can be deleted if you make this change:

```python3
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
        "protect_data": protect_data,
        "upv": protectserver,
        "snapshot_direct": entry.options.get(CONF_SNAPSHOT_DIRECT, False),
        "server_info": nvr_info,
        "disable_stream": entry.options.get(CONF_DISABLE_RTSP, False),
        "doorbell_text": entry.options.get(CONF_DOORBELL_TEXT, None),
    }
```
- All platforms

Remove commented code

`device_state_attributes` should be switched to `extra_state_attributes`

The attributes (ATTR_ATTRIBUTION) that are used on every entity could be moved to `extra_state_attributes`
in `entity.py` and all other platforms could do

```python3
    @property
    def extra_state_attributes(self):
        """Return the extra state attributes."""
        return {
            **super().extra_state_attributes,
            KEY: value, ....
        }
```

- `binary_sensor.py` 

Use `BinarySensorEntityDescription` (example in `tuya/binary_sensor.py)

- `camera.py`

The `ATTR_ONLINE` should be removed and the entity should be
`unavailable` instead when this is false via the `available` property

The `ATTR_LAST_TRIP_TIME` should be removed as its duplicated functionality
to the sensor entity and will cause additional db rows every time the
doorbell or motion sensor it tripped.

If there are any other dupicate `extra_state_attributes` they should
also be removed.

- `select.py`

Use `SelectEntityDescription` (example in `tuya/select.py`)

- `sensor.py`

Use `SensorEntityDescription`  (example in `tuya/sensor.py`)

- `number.py`

Use `NumberEntityDescription`  (example in `tuya/number.py`)

- `switch.py`

use `SwitchEntityDescription`  (example in `tuya/switch.py`)

- `light.py`

Nothing else stood out 👍

**Comment by briis** [comment]
Thanks Nick. I made 2 PR's for the `__init__.py` and all the Attributes. Now I am on the `XXXEntityDescription`. I looked at the Tuya files, and it is clear what this does, but I need an advice on how to implement this. The Tuya integration uses Discovery for finding the devices, whereas this integration, uses a set of criterias to add the entities and I am not sure how to go about this, as the `XXXEntityDescription` do not have fields for the extra stuff I need.

If we look at the `switch.py` as an example. I have the below list of entities to be added, but they should only be added if the device has the field `_SWITCH_REQUIRES`. As it is now, I pass this field in `SWITCH_TYPES` but this can not be part of the `SwitchEntityDescription`. How would you go about doing this the right way?

```python
_SWITCH_NAME = 0
_SWITCH_ICON = 1
_SWITCH_TYPE = 2
_SWITCH_REQUIRES = 3

SWITCH_TYPES = {
    "status_light": ["Status Light On", "led-on", "status_light", "has_ledstatus"],
    "hdr_mode": ["HDR Mode", "brightness-7", "hdr_mode", "has_hdr"],
    "high_fps": ["High FPS", "video-high-definition", "high_fps", "has_highfps"],
    "privacy_mode": ["Privacy Mode", "eye-settings", "privacy_mode", "privacy_on"],
}
```

**Comment by briis** [comment]
Spend a bit more time on this: Could I use a Mixin like below to extend the fields? Or is this the wrong way to go?

```python

@dataclass
class UnifiprotectRequiredKeysMixin:
    """Mixin for required keys."""

    ufp_max: int
    ufp_min: int
    ufp_step: int
    ufp_device_type: str
    ufp_required_field: str

@dataclass
class UnifiProtectNumberEntityDescription(
    NumberEntityDescription, UnifiprotectRequiredKeysMixin
):
    """Describes Unifi Protect Number entity."""


NUMBER_TYPES: tuple[UnifiProtectNumberEntityDescription, ...] = (
    UnifiProtectNumberEntityDescription(
        key=_ENTITY_WDR,
        name="Wide Dynamic Range",
        icon="mdi:state-machine",
        entity_category=ENTITY_CATEGORY_CONFIG,
        ufp_min=0,
        ufp_max=3,
        ufp_step=1,
        ufp_device_type=DEVICES_WITH_CAMERA,
        ufp_required_field=None,
    ),
    UnifiProtectNumberEntityDescription(
        key=_ENTITY_MIC_LEVEL,
        name="Microphone Level",
        icon="mdi:microphone",
        entity_category=ENTITY_CATEGORY_CONFIG,
        ufp_min=0,
        ufp_max=100,
        ufp_step=1,
        ufp_device_type=DEVICES_WITH_CAMERA,
        ufp_required_field=None,
    ),
    UnifiProtectNumberEntityDescription(
        key=_ENTITY_ZOOM_POS,
        name="Zoom Position",
        icon="mdi:magnify-plus-outline",
        entity_category=ENTITY_CATEGORY_CONFIG,
        ufp_min=0,
        ufp_max=100,
        ufp_step=1,
        ufp_device_type=DEVICES_WITH_CAMERA,
        ufp_required_field="has_opticalzoom",
    ),
)
````

**Comment by bdraco** [comment]
A mixin willl work.  There are some examples in `august/binary_sensor.py` or `august/sensor.py`

Also There are more select examples in `xiaomi_miio/select.py`

**Comment by AngellusMortis** [comment]
Going to go a head and close this since we have a path forward/roadmap on where to go.
